internal/scylla/client.go

```diff
@@ -136,11 +136,65 @@ func (c *Client) doHttpRequest(req *http.Request) (resp *http.Response, temporar
 		return
 	}
 
-	temporaryErr = resp.StatusCode == http.StatusBadGateway || resp.StatusCode == http.StatusGatewayTimeout || resp.StatusCode == http.StatusTooManyRequests
+	body, err := getBody(resp)
+	if err != nil {
+		return nil, false, fmt.Errorf("failed to get response body: %w", err)
+	}
+
+	body2, err := getBody(resp)
+	if err != nil {
+		return nil, false, fmt.Errorf("failed to get response body: %w", err)
+	}
+
+	code := getErrorCode(body)
+
+	tflog.Trace(req.Context(), "getErrorCode: "+code, map[string]interface{}{
+		"code":   resp.StatusCode,
+		"status": resp.Status,
+		"body1":  string(body),
+		"body2":  string(body2),
+	})
+
+	if code == "000001" {
+		return resp, true, fmt.Errorf("too many request response")
+	}
+	temporaryErr = temporaryErr || resp.StatusCode == http.StatusBadGateway || resp.StatusCode == http.StatusGatewayTimeout || resp.StatusCode == http.StatusTooManyRequests
 	return
 }
 
+func getBody(resp *http.Response) ([]byte, error) {
+	if resp.Body == nil || resp.Body == http.NoBody {
+		return nil, nil
+	}
+	defer resp.Body.Close()
+	var buf bytes.Buffer
+	if _, err := buf.ReadFrom(resp.Body); err != nil {
+		return nil, err
+	}
+	resp.Body = io.NopCloser(bytes.NewReader(buf.Bytes()))
+	return buf.Bytes(), nil
+}
+
+func getErrorCode(body []byte) string {
+	if len(body) == 0 {
+		return ""
+	}
+
+	var data = struct {
+		Error string `json:"error"`
+	}{}
+
+	if err := json.Unmarshal(body, &data); err != nil {
+		// Not every response carry json, so we return empty string here.
+		return ""
+	}
+	return data.Error
+}
 func (c *Client) doHttpRequestWithRetries(req *http.Request, retries int, retryBackoffDuration time.Duration) (*http.Response, error) {
+	reqCpy, err := cloneRequest(req)
```

Comment:
1. the implementation of cloneRequest is wrong, you set req.Body without setting req.GetBody func
2. you don't need to clone the req to make it retryable

Reply:
Please don't copy the request, there is no such need. I don't understand why it would look more confusing, could you please elaborate? The standard library api `net/http` is designed to modify `http.Request` in-place, given that I don't understand why anyone should be "confused"...

Reply:
Also if you read the whole body, please use some io.LimitReader with high limit like 1MiB. Just to prevent any rogue payloads.

Reply:
> Please don't copy the request, there is no such need. I don't understand why it would look more confusing, could you please elaborate? The standard library api `net/http` is designed to modify `http.Request` in-place, given that I don't understand why anyone should be "confused"...

Since `doHttpRequestWithRetries` is calling `c.doHttpRequest(req)` which reads `req.Body` in order to rewind `body` you need to store it somwhere before it is passed to `c.doHttpRequest(req)`.

So, there are two ways to solve it:
1. Store body in copy of request, current solution
2. Store body in some variable and push it back to same req:
```
func (c *Client) doHttpRequestWithRetries(req *http.Request, retries int, retryBackoffDuration time.Duration) (*http.Response, error) {
	bodyCopy, err := getRequestBody(req)
	if err != nil {
		return nil, fmt.Errorf("failed to get request body: %w", err)
	}
	resp, temporaryErr, err := c.doHttpRequest(req)
	if temporaryErr && retries > 0 {
		if err == nil {
			_ = resp.Body.Close() // We want to retry anyway.
		}
		var timeToSleep time.Duration
		if d, ok := parseRetryAfter(resp.Header.Get("Retry-After")); ok {
			timeToSleep = d
		} else {
			timeToSleep = retryBackoffDuration
		}
		timer := time.NewTimer(timeToSleep)
		defer timer.Stop()
		select {
		case <-timer.C:
		case <-req.Context().Done():
			return nil, req.Context().Err()
		}
                
                req.Body = bodyCopy
		return c.doHttpRequestWithRetries(req, retries-1, retryBackoffDuration*2)
	}

	return resp, err
}
```

Which to mee looks more confusing that just having a copy of the request.

Reply:
> Also if you read the whole body, please use some io.LimitReader with high limit like 1MiB. Just to prevent any rogue payloads.

Not quite sure if I understand it correctly ? Do you want to send only 1MiB of data on retry ?

Reply:
@dkropachev See my explanation here: https://github.com/scylladb/terraform-provider-scylladbcloud/pull/100#discussion_r1512231778

Reply:
@rjeczalik , please see my [response](https://github.com/scylladb/terraform-provider-scylladbcloud/pull/100#discussion_r1512632759) and also take a look at the code, I have updated it to rewind body in place.